src/query/block/reconsolidated.go

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package block
+
+import (
+	"time"
+
+	"github.com/m3db/m3/src/query/ts"
+)
+
+type reconsolidatedBlock struct {
+	block Block
+}
+
+// NewReconsolidatedBlock creates a lazy block wrapping another block with lazy options.
+func NewReconsolidatedBlock(block Block) Block {
+	return &reconsolidatedBlock{
+		block: block,
+	}
+}
+
+func (b *reconsolidatedBlock) Close() error { return b.block.Close() }
+
+func (b *reconsolidatedBlock) WithMetadata(
+	meta Metadata,
+	sm []SeriesMeta,
+) (Block, error) {
+	bl, err := b.block.WithMetadata(meta, sm)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewReconsolidatedBlock(bl), nil
+}
+
+// StepIter returns a StepIterator
+func (b *reconsolidatedBlock) StepIter() (StepIter, error) {
+	return b.block.StepIter()
+}
+
+// SeriesIter returns a SeriesIterator
+func (b *reconsolidatedBlock) SeriesIter() (SeriesIter, error) {
+	return b.block.SeriesIter()
+}
+
+// Unconsolidated returns the unconsolidated version for the block
+func (b *reconsolidatedBlock) Unconsolidated() (UnconsolidatedBlock, error) {
```

Comment:
Hm, wonder if we should come up with some terms for this.  `reconsolidatedBlock` and `ucReconsolidatedBlock` definitely took me a minute or so to totally grok the relationship, etc just based on name.  I get it know, but yeah, hehe.